### PR TITLE
Add lambda support

### PR DESCRIPTION
**Pull request by tylersouthwick**

### CODE REVIEW

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/IAM.scala [hunk 1]
```diff
@@ -88,9 +88,11 @@ case class `AWS::IAM::Role`(
   Path:                     Option[Token[String]] = None,
   Policies:                 Option[Seq[Policy]] = None,
   override val Condition: Option[ConditionRef] = None
-  ) extends Resource[`AWS::IAM::Role`]{
+  ) extends Resource[`AWS::IAM::Role`] with HasArn {
```
Comment (ddgenome) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/IAM.scala]: A left over from your previous PR?

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    override def read(json: JsValue): Runtime = json match {
+      case JsString(runtime) => new Runtime(runtime)
+    }
+  }
+
+}
+
+case class `AWS::Lambda::Function`(name: String,
+                                   Code: Code,
+                                   Description: String,
```
Comment (ddgenome) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala]: Shouldn't Description be an `Option[Token[String]]`?

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  def when(newCondition: Option[ConditionRef] = Condition) = copy(Condition = newCondition)
+}
+
+object `AWS::Lambda::Permission` extends DefaultJsonProtocol {
+  implicit val format: JsonFormat[`AWS::Lambda::Permission`] = jsonFormat7(`AWS::Lambda::Permission`.apply)
+}
+
+case class `AWS::Lambda::EventSourceMapping`(
+                                              name: String,
+                                              BatchSize: Token[Int],
```
Comment (ddgenome) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala]: `BatchSize: Option[Token[Int]] = None`

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                                              Enabled: Token[Boolean],
```
Comment (ddgenome) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala]: `Enabled: Option[Token[Boolean]] = None`

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                                              EventSourceArn: Token[String],
+                                              FunctionName: Token[String],
```
Comment (ddgenome) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala]: Perhaps FunctionName should be a ```Token[ResourceRef[`AWS::Lambda::Function`]]```?

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                S3ObjectVersion: Option[Token[String]],
+                ZipFile: Option[String])
+
+object Code extends DefaultJsonProtocol {
+  implicit val format: JsonFormat[Code] = jsonFormat4(Code.apply)
+}
+
+case class `AWS::Lambda::Permission`(name: String,
+                                     Action: String,
+                                     FunctionName: Token[String],
```
Comment (ddgenome) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala]: Perhaps FunctionName should be a Token[ResourceRef[`AWS::Lambda::Function`]]?

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                                     Principal: Token[String],
+                                     SourceAccount: Option[Token[String]],
+                                     SourceArn: Token[String],
```
Comment (ddgenome) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala]: `SourceArn: Option[Token[String]]`

**Comment by ddgenome** [comment]
Thank you very much for this, and all your other contributions!

Overall the code looks very good. I've made a few minor comments that we'd love to hear your thoughts on.

Eventually, it seems we would like to address the fact that the S3 and ZipFile options in Code are mutually exclusive, and that ZipFile is only an option with the nodejs RunTime. We may also want to start thinking about modeling ARNs in some way.

Thanks again!

**Comment by tylersouthwick** [comment]
Yeah, I was going back and forth about the s3 and zipFile. It depends on how far from the cloudformation 1-1 primitives should this library go?

**Comment by ddgenome** [comment]
The primitives can be 1:1, but there are a few ways we make them exclusionary either at compile or run time.

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+object `AWS::Lambda::Permission` extends DefaultJsonProtocol {
+  implicit val format: JsonFormat[`AWS::Lambda::Permission`] = jsonFormat7(`AWS::Lambda::Permission`.apply)
+}
+
+case class `AWS::Lambda::EventSourceMapping`(
+                                              name: String,
+                                              BatchSize: Option[Token[Int]],
+                                              Enabled: Option[Token[Boolean]],
+                                              EventSourceArn: Token[String],
+                                              FunctionName: Token[String],
```
Comment (ddgenome) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala]: This function name should probably be ```Token[ResourceRef[`AWS::Lambda::Function`]]``` too.

**Comment by ddgenome** [comment]
@tylersouthwick can you change that second FunctionName type to ```Token[ResourceRef[`AWS::Lambda::Function`]]``` so we can get this merged in? Thanks again!

**Comment by tylersouthwick** [comment]
both FunctionName paramters are Token[ResourceRef[`AWS::Lambda::Function`]]

**Comment by ddgenome** [comment]
Thanks @tylersouthwick . Perhaps @tj-corrigan can merge this in?

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala [hunk 9]
```diff
@@ -0,0 +1,92 @@
+package com.monsanto.arch.cloudformation.model.resource
+
+import com.monsanto.arch.cloudformation.model._
+import spray.json.{JsString, JsValue, JsonFormat, DefaultJsonProtocol}
+
+class Runtime(val runtime: String)
```
Comment (kevinmeredith) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/Lambda.scala]: I have not used Lambda for almost 1 year. 

However, if we know the universe of `runtime` values, then I believe using an [algrebraic data type](http://stackoverflow.com/questions/26689951/writing-algebraic-data-type-in-scala) should be considered.

```
sealed trait Runtime
case object Java8 extends Runtime
case object ...
```

**Comment by tj-corrigan** [comment]
thanks for the PR, I've pushed out a new version (3.0.5)

**Pull request merged**